### PR TITLE
Implement abstraction for Encryption module

### LINKED ISSUE #15: Implement Encryption module
ℹ️ This issue is part of the Epic #16

### Describe the solution you'd like

Implement a module to allow encryption and secrets generation

### Describe alternatives you've considered

Implement Erlang's Crypto dependency
Define abstraction for Encryption

### PR DESCRIPTION
**Pull request by miguelnietoa**
## Description

This PR adds an abstraction for the Encryption module.

Fixes #15 

## Type of change
- [x] New feature (non-breaking change which adds functionality)

## Checklist

- [x] My code follows the style guidelines of this project
- [x] I have performed a self-review of my code
- [x] I have commented my code, particularly in hard-to-understand areas
- [x] I have made corresponding changes to the documentation
- [x] My changes generate no new warnings
- [x] I have added tests that prove my fix is effective or that my feature works
- [x] New and existing unit tests pass locally with my changes

### CODE REVIEW

File: apps/mintacoin/lib/mintacoin/encryption/default.ex [hunk 1]
```diff
@@ -0,0 +1,140 @@
...
+
+    ciphertext =
+      @cipher
+      |> :crypto.crypto_one_time(secret_key, iv, plaintext, true)
+      |> (&(iv <> &1)).()
+      |> Base.encode64(padding: false)
+
+    {:ok, ciphertext}
+  rescue
+    :error -> {:error, :decode_64_error}
```
Comment (juanhurtado10) [inline apps/mintacoin/lib/mintacoin/encryption/default.ex:46]: @miguelnietoa what do you think about using more readable error messages like, `:decoding_error, :encryption_error`?

File: apps/mintacoin/lib/mintacoin/encryption/default.ex [hunk 2]
```diff
@@ -0,0 +1,140 @@
...
+      {:ok, plaintext}
+    else
+      :error -> {:error, :decode_64_error}
+      _error -> {:error, :error_in_decryption_pattern_matching}
+    end
+  end
+
+  @impl true
+  def random_keypair do
+    :eddsa
```
Comment (juanhurtado10) [inline apps/mintacoin/lib/mintacoin/encryption/default.ex:68]: @miguelnietoa @peterchicaiza please, use a constant.

File: apps/mintacoin/lib/mintacoin/encryption/default.ex [hunk 3]
```diff
@@ -0,0 +1,140 @@
...
+    :binary.part(data, 0, byte_size(data) - to_remove)
+  end
+
+  @spec pad(data :: String.t(), block_size :: integer()) :: String.t()
+  defp pad(data, block_size) do
+    to_add = block_size - rem(byte_size(data), block_size)
+    data <> :binary.copy(<<to_add>>, to_add)
+  end
+
+  @spec encode_keypair({public_key :: String.t(), secret_key :: String.t()}) ::
```
Comment (juanhurtado10) [inline apps/mintacoin/lib/mintacoin/encryption/default.ex:132]: @peterchicaiza @miguelnietoa 

1. a better way to write this spec down.

```elixir
@spec encode_keypair(key_pair :: key_pair()) :: {:ok, key_pair()} 
``` 

2. IMO, this function will never return a `{:error, String.t()}`

File: apps/mintacoin/lib/mintacoin/encryption/default.ex [hunk 4]
```diff
@@ -0,0 +1,140 @@
...
+
+  @impl true
+  def random_keypair do
+    :eddsa
+    |> :crypto.generate_key(:ed25519)
+    |> encode_keypair()
+  end
+
+  @impl true
+  def pk_from_sk(secret_key_64) do
```
Comment (juanhurtado10) [inline apps/mintacoin/lib/mintacoin/encryption/default.ex:74]: @peterchicaiza @miguelnietoa IMO, using acronyms generates confusion.

File: apps/mintacoin/lib/mintacoin/encryption/default.ex [hunk 5]
```diff
@@ -0,0 +1,140 @@
...
+    {:ok, secret_key} = Base.decode64(secret_key_64, padding: false)
+
+    :eddsa
```
Comment (juanhurtado10) [inline apps/mintacoin/lib/mintacoin/encryption/default.ex:77]: @miguelnietoa @peterchicaiza Please, use a constant

File: apps/mintacoin/mix.exs [hunk 6]
```diff
@@ -41,7 +41,8 @@ defmodule Mintacoin.MixProject do
       {:ecto_sql, "~> 3.6"},
       {:postgrex, ">= 0.0.0"},
       {:jason, "~> 1.2"},
-      {:swoosh, "~> 1.3"}
+      {:swoosh, "~> 1.3"},
```
Comment (juanhurtado10) [inline apps/mintacoin/mix.exs:44]: @peterchicaiza @miguelnietoa This dependency is not really needed

File: apps/mintacoin/test/mintacoin/encryption_test.exs [hunk 7]
```diff
@@ -0,0 +1,93 @@
...
+      {:error, :error_in_encryption} = Encryption.encrypt("test", "invalid")
+    end
+  end
+
+  describe "decrypt/2" do
+    test "with valid secret should return a plaintext" do
+      secret = Encryption.generate_secret()
+      {:ok, ciphertext} = Encryption.encrypt("test", secret)
+      {:ok, plaintext} = Encryption.decrypt(ciphertext, secret)
+      "test" = plaintext
```
Comment (juanhurtado10) [inline apps/mintacoin/test/mintacoin/encryption_test.exs:44]: @miguelnietoa @peterchicaiza Why not just

```elixir
 {:ok, "test"} = Encryption.decrypt(ciphertext, secret)
```

File: apps/mintacoin/test/mintacoin/encryption_test.exs [hunk 8]
```diff
@@ -0,0 +1,93 @@
...
+      {:ok, {public_key, secret_key}} = Encryption.random_keypair()
+      refute is_nil(public_key)
+      refute is_nil(secret_key)
+    end
+  end
+
+  describe "pk_from_sk/1" do
+    test "with a valid secret key, it should return the keypair with the right public key" do
+      {:ok, {pk, sk}} = Encryption.random_keypair()
+      {:ok, {assumed_pk, assumed_sk}} = Encryption.pk_from_sk(sk)
```
Comment (juanhurtado10) [inline apps/mintacoin/test/mintacoin/encryption_test.exs:63]: @peterchicaiza @miguelnietoa a better assert would be

```elixir
{:ok, {^pk, ^sk}} = Encryption.pk_from_sk(sk)
```

File: apps/mintacoin/test/mintacoin/encryption_test.exs [hunk 9]
```diff
@@ -0,0 +1,93 @@
...
+      refute is_nil(Encryption.generate_secret())
+    end
+  end
+
+  describe "one_time_token/0" do
+    test "Should return the encoded, and the hashed token" do
+      {:ok, {encoded_token, hashed_token}} = Encryption.one_time_token()
+
+      ^hashed_token =
+        encoded_token
```
Comment (juanhurtado10) [inline apps/mintacoin/test/mintacoin/encryption_test.exs:21]: @peterchicaiza @miguelnietoa use setup to perform this operation instead

**Review by juanhurtado10 — CHANGES_REQUESTED** [review]

File: apps/mintacoin/lib/mintacoin/encryption/default.ex (hunk 1 above)
Comment (miguelnietoa) [reply apps/mintacoin/lib/mintacoin/encryption/default.ex:46]: Sure!

**Comment by miguelnietoa** [comment]
Hi there, @juanhurtado10 

We applied the suggested changes.
To facilitate the code review process, we have planned to [remove the Swoosh dependency in another PR](https://github.com/kommitters/mintacoin/issues/32). As well as [separating some functions within the KeyPair module](https://github.com/kommitters/mintacoin/issues/33).

File: apps/mintacoin/mix.exs (hunk 6 above)
Comment (miguelnietoa) [reply apps/mintacoin/mix.exs:44]: https://github.com/kommitters/mintacoin/issues/32

**Review by juanhurtado10 — APPROVED** [review]
@miguelnietoa @peterchicaiza Awesome job done 👏  🚀

**Pull request merged**